### PR TITLE
pay: Remember and update channel_hints across payments

### PR DESCRIPTION
**Pull request by cdecker**
Builds on top of #7487.

This PR takes the `channel_hint`s that were previously created,
updated and then discarded as part of a payment, and instead we are
making them long-lived. This means that we share what we learn from
one payment with all later payments, allowing us to skip attempts that
will likely fail anyway, and improve the overall performance
(time-to-completion as well as success rate of payments).

The downside of this is that we are now keeping observations for
prolonged periods, meaning that these observations are no longer valid
when we use them. If used unchanged these earlier observations could
lead us to skip routes that we deem unavailable due to a prior
observation, but which is now back to being usable, which in turn
impacts our ability to complete a payment successfully. The core
contribution in this PR therefore is the time-based relaxation of
learnt constraints. It is based on a leaky-bucket approach, with a
linear refill-rate for depleted channels, such that after 2h we
consider any observation outdated.

## Example

If at time `t` we observe a channel `c` with `cap_c` as overall
capacity, and `bal_c = 0.5 * cap_c` the currently available balance
(i.e., we observed a payment of `bal_c + 1` fail to generate this
hint), then we have a refill rate `r = cap_c / 7200`, and after 1h the
channel is considered to be fully available again, i.e., we don't have
a more restrictive observation, and it is old enough that it does not
add to the structural information we already have from the gossip.

```mermaid
graph TD;
    CLN7487 --> CLN7494 ;
    click CLN7487 "https://github.com/ElementsProject/lightning/pull/7487"
    click CLN7494 "https://github.com/ElementsProject/lightning/pull/7494"
    style CLN7494 fill:#AAFFAA
```

### CODE REVIEW

File: common/route.h [hunk 1]
```diff
@@ -18,13 +18,15 @@ struct gossmap_node;
  * @direction: 0 (dest node_id < src node_id), 1 (dest node_id > src).
  * @node_id: the node_id of the destination of this hop.
  * @amount: amount to send through this hop.
+ * @total_amount: The amount the channel was funded with.
```
Comment (rustyrussell) [inline common/route.h:21]: I would actually just use htlc_max, but if you want this, please call it "capacity"?

File: common/route.c [hunk 2]
```diff
@@ -113,6 +109,8 @@ static bool dijkstra_to_hops(struct route_hop **hops,
 	if (!dijkstra_to_hops(hops, gossmap, dij, next, amount, cltv))
 		return false;
 
+	gossmap_chan_get_capacity(gossmap, c, &total);
+	(*hops)[num_hops].total_amount.millisatoshis = total.satoshis * 1000; /* Raw: simpler. */
```
Comment (rustyrussell) [inline common/route.c:113]: Please use amount_sat_to_msat().  I'm happy for you to abort or return false, but this just silently overflows.

File: lightningd/onchain_control.c [hunk 3]
```diff
@@ -518,6 +518,44 @@ static u32 infinite_block_deadline(const struct chain_topology *topo)
 	return get_block_height(topo) + 300;
 }
 
+/**
+ * slow_sweep_dealine -- A smart(-ish) way of estimating sweep fees.
+ *
+ * We use the `close_blockheight` as the basis for the target. We then
+ * target to have the funds back at most 2 weeks after the close,
```
Comment (rustyrussell) [inline lightningd/onchain_control.c:525]: This final commit does not belong in this PR?

**Review by rustyrussell — CHANGES_REQUESTED** [review]

**Comment by rustyrussell** [comment]
I rebased, removed the final commit (which was from a different PR, and already merged) and added a commit with my suggested changes.

But I'm not sure this PR is complete?

File: plugins/channel_hint.c [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+	refill = amount_msat_div(refill, PAY_REFILL_TIME);
+	if (!amount_msat_add(&hint->estimated_capacity,
+			     hint->estimated_capacity, refill))
+		abort();
+
+	/* Clamp the value to the `overall_capacity` */
+	if (amount_msat_greater(hint->estimated_capacity,
+				hint->overall_capacity))
+		hint->estimated_capacity = hint->overall_capacity;
```
Comment (Lagrang3) [inline plugins/channel_hint.c:53]: I think it is unlikely to hit an overflow under normal conditions. For example a timeframe of 1 year ~3e7 seconds times a channel capacity of 1 BTC = 1e11 msat, when multiplied gives ~3e18 that fits into 64 bits.

I would anyways do the computation only if `seconds` is below `PAY_REFILL_TIME` 'cause otherwise we know immediately that the estimated capacity has been refilled completely:
```suggestion
	if (seconds >= PAY_REFILL_TIME) {
		/* no computation is needed, no overflow */
		hint->estimated_capacity = hint->overall_capacity;
	} else {
		if (!amount_msat_mul(&refill, hint->overall_capacity, seconds))
			abort();

		refill = amount_msat_div(refill, PAY_REFILL_TIME);
		if (!amount_msat_add(&hint->estimated_capacity,
				     hint->estimated_capacity, refill))
			abort();

		/* Clamp the value to the `overall_capacity` */
		hint->estimated_capacity = amount_msat_min(
		    hint->overall_capacity, hint->estimated_capacity);
	}
```

File: common/route.h (hunk 1 above)
Comment (cdecker) [reply common/route.h]: No, this needs to be the overall capacity (yes, I'll rename it), since the HTLC limit is, as the name suggests, per HTLC, and we are running a tally of how much we've already assigned to the channel.

File: lightningd/onchain_control.c (hunk 3 above)
Comment (cdecker) [reply lightningd/onchain_control.c:525]: The rebase removed it, because the onchaind PR already made it into `master`.

File: plugins/channel_hint.c (hunk 4 above)
Comment (cdecker) [reply plugins/channel_hint.c:52]: I don't think there is a logical error in the current code? Are you mostly looking at optimizations? In that case I'd defer to a separate PR, since it is not wrong per se, maybe suboptimal.

**Comment by cdecker** [comment]
Ok, I checked this once more, and decided to draft it quickly, fix it up, and then undraft as soon as possible. I noticed that while we were creating a global hint set, we didn't use it. So this is just me making that work.

**Comment by cdecker** [comment]
@rustyrussell can you help me with the sat to msat conversion? I seem to get confused by the dereferencing construction there, and as soon as I reintroduce the `amount_sat_to_msat()` call I get the following `valgrind` error:

```
==37765== Conditional jump or move depends on uninitialised value(s)
==37765==    at 0x136050: amount_sat_to_msat (amount.c:15)
==37765==    by 0x10CCBE: dijkstra_to_hops (route.c:114)
==37765==    by 0x10CE40: route_from_dijkstra (route.c:138)
==37765==    by 0x120725: try_route (topology.c:123)
==37765==    by 0x120A2D: listpeerchannels_getroute_done (topology.c:169)
==37765==    by 0x12546C: handle_rpc_reply (libplugin.c:1016)
==37765==    by 0x1261CE: rpc_read_response_one (libplugin.c:1202)
==37765==    by 0x126351: rpc_conn_read_response (libplugin.c:1226)
==37765==    by 0x275DA0: next_plan (io.c:60)
==37765==    by 0x2769F6: do_plan (io.c:422)
==37765==    by 0x276A5D: io_ready (io.c:439)
==37765==    by 0x278E91: io_loop (poll.c:455)
==37765== 
```

That indirection is rather messy, and I wonder if I'm accidentally accessing memory next to the value I actually want to set :shrug: 

Anyway, with the next PR in the series I am reworking the set of channel hints to be its own entity, and encapsulate all the operations on sets of channel hints, which should also free us from the pointer magic.

**Comment by rustyrussell** [comment]
I've moved this to the next milestone.   I should probably have done that for the prior PR, which is now kind of useless for this release :(

The approach here contrasts with askrene, which uses explicit calls to inform about everything it learns (rather than notifications).  But it's probably better that they remain separate for now (the redundancy kind of bugs me though!).

**Comment by cdecker** [comment]
> I've moved this to the next milestone. I should probably have done that for the prior PR, which is now kind of useless for this release :(

Then revert it, I just tried my best to get it into the release, and didn't manage in time.

> The approach here contrasts with askrene, which uses explicit calls to inform about everything it learns (rather than notifications). But it's probably better that they remain separate for now (the redundancy kind of bugs me though!).

Generally speaking we lack direction in all of these forays. askrene and renepay are unusable in their current state, and we have loads of users complaining about the `pay` performance, without being helpful ("stuff doesn't work" is not actionable feedback!). We wanted to replace `pay` with `renepay` so this is meant just as a way to explore data sharing. I very much dislike the tight coupling via RPC methods, not to speak, exposing internals via the RPC polluting out docs is not good either (it confuses users, and makes internals public and you can bet users will start relying on them...). Thoroughly not enjoying any of this!

File: common/route.c [hunk 5]
```diff
@@ -94,12 +94,7 @@ static bool dijkstra_to_hops(struct route_hop **hops,
 	/* OK, populate other fields. */
 	c = dijkstra_best_chan(dij, curidx);
 
-	if (c->half[0].nodeidx == curidx) {
-		(*hops)[num_hops].direction = 0;
-	} else {
-		assert(c->half[1].nodeidx == curidx);
-		(*hops)[num_hops].direction = 1;
-	}
+	(*hops)[num_hops].direction = c->half[0].nodeidx == curidx ? 0 : 1;
```
Comment (rustyrussell) [inline common/route.c:97]: This removes the assertion that curidx is indeed one side of the channel...

File: common/json_stream.c [hunk 6]
```diff
@@ -626,6 +626,14 @@ void json_add_amount_msat(struct json_stream *result,
 	json_add_u64(result, msatfieldname, msat.millisatoshis); /* Raw: low-level helper */
 }
 
+void json_add_amount_sat(struct json_stream *result,
+			  const char *satfieldname,
+			  struct amount_sat sat)
+{
+	assert(strends(satfieldname, "_sat") || streq(satfieldname, "sat"));
+	json_add_u64(result, satfieldname, sat.satoshis); /* Raw: low-level helper */
```
Comment (rustyrussell) [inline common/json_stream.c:634]: Disagree.  There's a reason this function was removed.  In general, we want to expose values in msat.  In this case, the fact that the capacity has to be in whole sats is kind of a detail.

File: tests/test_connection.py [hunk 7]
```diff
@@ -4368,49 +4368,6 @@ def test_multichan(node_factory, executor, bitcoind):
...
-                                              'dev-no-reconnect': None})
-
-    def send_many_payments():
-        for i in range(20):
-            time.sleep(0.5)
-            inv = l2.rpc.invoice(100, "label-" + str(i), "desc")['bolt11']
-            try:
-                l1.rpc.pay(inv)
-            except RpcError:
-                pass
```
Comment (rustyrussell) [inline tests/test_connection.py:4383]: You didn't read this carefully enough.  It's ok if the payments fail!

But it's here because it ticked a reconnection desync when payments happened.  If your pay modifications are interfering with this test, they probably need to not create hints on local channels?

**Comment by cdecker** [comment]
Ok, I think I addressed all the feedback:

 - The channel_hint total_amount is now in msat again, saving us a couple of conversions
 - The `test_mutual_reconnect_race` test has been re-added with a step down so a failed payment doesn't immediately cause all others to fail

Can we merge this now?
